designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+
+The Stan CSV file, however, contains far more information than a single table's worth of data.
+To cram all the information from a CmdStan run we 
+abuse the CSV format by using blocks of comment lines to record additional information.
+The current Stan API requires the information in a comment line be assembled and flattened into a text string.
+Even when we can use a fast CSV parser to extract the CSV data, we still need to build ad-hoc parsers for the comment blocks
+to re-assemble the flattened information.
+
+The sample methods produce the largest amount of output data and these are the most difficult to process quickly and efficiently.
+The first part of a CmdStan CSV file always a series of comment rows which contain configuration and version information,
+and this comment header continues to accumulate new bits of information about program version and compilation options.
+The NUTS-HMC methods output a series of comment lines which contain the stepsize and metric.
+The sample methods also record timing information at the end.
+
+In order to monitor the progress of CmdStan's inference algorithms,
+the outputs from CmdStan must be available to downstream readers on a streaming basis.
+To facilitate monitoring different aspects of the algorithm, a more granular set of outputs,
+instead of a single CSV file would make this task easier.
+Another consideration is the use of multi-chain methods; currently multi-chain NUTS-HMC and in the future,
+multi-chain Pathfinder.  When running any kind of ensemble method, we need to record the process/chain_id
+and make this available 
+
+Therefore we need to re-implement the set of output methods used by CmdStan and the core Stan services
+to capture more information in a more structured way.
+At the same time, we need to continue to support the current Stan CSV format in order to support existing downstream processing packages.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For a given run of an inference algorithm given a model and data,
+the outputs of interest can be categorized in terms of information source and content
+and information structure.
+
+There are three sources of information:
+
+1. Stan program outputs: estimates for all model parameters, transformed parameters, and generated quantities.
+
+2. Inference engine outputs: various diagnostics.
+For the NUTS-HMC sampler, the current outputs are the stepsize and metric, and the per-iteration sampler variables (state).
+The optimization and variational algorithms also report on their respective iterations.
+
+3. Interface-level outputs:
+   + model name, compile options, Stan compiler and (Cmd)Stan version and compile options.
+   + inference algorithm, user-specified configuration options, and default config.
+   + input and output data descriptors.
+   + process-level information:  chain, iteration.
+   + timing information: timestamp information reported by processing events.
```

Comment:
Do we really want to save timing info?

Timestamping (when it was run) is more relevant.

Reply:
no, we don't want timing info - as you say - timestamping much more useful.